// @jsse/pbfont/dist/index.js
// src/index.ts
import { create, fromBinary, toBinary, toJsonString } from "@bufbuild/protobuf";

// src/gen/glyphs_pb.ts
import { fileDesc, messageDesc } from "@bufbuild/protobuf/codegenv1";
var file_glyphs = /* @__PURE__ */ fileDesc(
  "CgxnbHlwaHMucHJvdG8SC2xsbXIuZ2x5cGhzIm4KBWdseXBoEgoKAmlkGAEgAigNEg4KBmJpdG1hcBgCIAEoDBINCgV3aWR0aBgDIAIoDRIOCgZoZWlnaHQYBCACKA0SDAoEbGVmdBgFIAIoERILCgN0b3AYBiACKBESDwoHYWR2YW5jZRgHIAIoDSJMCglmb250c3RhY2sSDAoEbmFtZRgBIAIoCRINCgVyYW5nZRgCIAIoCRIiCgZnbHlwaHMYAyADKAsyEi5sbG1yLmdseXBocy5nbHlwaCI3CgZnbHlwaHMSJgoGc3RhY2tzGAEgAygLMhYubGxtci5nbHlwaHMuZm9udHN0YWNrKgUIEBCAQEICSAM"
);
var glyphSchema = /* @__PURE__ */ messageDesc(file_glyphs, 0);
var fontstackSchema = /* @__PURE__ */ messageDesc(file_glyphs, 1);
var glyphsSchema = /* @__PURE__ */ messageDesc(file_glyphs, 2);

// src/index.ts
function debug(buffer) {
  if (buffer instanceof Uint8Array) {
    const g = fromBinary(glyphsSchema, buffer);
    return toJsonString(glyphsSchema, g);
  }
  return toJsonString(glyphsSchema, buffer);
}
function decode(buffer) {
  return fromBinary(glyphsSchema, buffer);
}
function encode(data) {
  return toBinary(glyphsSchema, data);
}
function range256(start) {
  if (start < 0 || start > 65535)
    throw new Error(`start must be between 0 and 255; given ${start}`);
  const start256 = Math.trunc(start / 256) * 256;
  const stop256 = start256 + 255;
  return {
    start: start256,
    stop: stop256,
    str: `${start256}-${stop256}`
  };
}
function parseRange(range) {
  const [start, stop] = range.split("-").map(Number);
  if (start === void 0 || stop === void 0 || Number.isNaN(start) || Number.isNaN(stop) || start < 0 || stop < 0 || start > 65535 || stop > 65535) {
    throw new Error(`range must be in the form 'start-stop'; given ${range}`);
  }
  return { start, stop, str: `${start}-${stop}` };
}
function combineRanges(ranges) {
  const parsedRanges = ranges.map((element) => parseRange(element));
  const start = Math.min(...parsedRanges.map((r) => r.start));
  const stop = Math.max(...parsedRanges.map((r) => r.stop));
  return { start, stop, str: `${start}-${stop}` };
}
function combine(buffers, fontstackName) {
  if (buffers.length === 0) return;
  if (buffers.length === 1) return buffers[0];
  const combinedGlyphs = [];
  const coverage = /* @__PURE__ */ new Set();
  const names = [];
  const rangeStrings = [];
  for (const buffer of buffers) {
    const {
      stacks: [currentFontstack]
    } = decode(buffer);
    const {
      glyphs: currentGlyphs = [],
      range: fsRange,
      name: currentFontstackName
    } = currentFontstack || {};
    if (fsRange) rangeStrings.push(fsRange);
    if (currentFontstackName) names.push(currentFontstackName);
    for (const glyph of currentGlyphs) {
      if (!coverage.has(glyph.id)) {
        combinedGlyphs.push(glyph);
        coverage.add(glyph.id);
      }
    }
  }
  combinedGlyphs.sort((a, b) => a.id - b.id);
  const range = rangeStrings.length > 0 ? combineRanges(rangeStrings) : range256(Math.min(...coverage));
  const { str: rangeStr } = range;
  const resultFontstack = create(fontstackSchema, {
    name: fontstackName || names.join(", "),
    glyphs: combinedGlyphs,
    range: rangeStr || ""
  });
  return encode(create(glyphsSchema, { stacks: [resultFontstack] }));
}
export {
  combine,
  debug,
  decode,
  encode,
  file_glyphs,
  fontstackSchema,
  glyphSchema,
  glyphsSchema
};
